File: NeetCode/Backtracking/Generate_Paranthese/Generate_Paranthese.cpp
// Generate_Paranthese.cpp
//
// Compile locally with:
// clang++ Generate_Paranthese.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<string> generateParenthesis(int n)
    {
        string s = "";
        vector<string> results;
        dfs(s, 0, 0, results, n);
        return results;
    }

    void dfs(string s, int left, int right, vector<string> &results, int n)
    {
        if (s.size() == 2 * n)
        {
            results.push_back(s);
        }

        if (left < n)
        {
            dfs(s + "(", left + 1, right, results, n);
        }
        if (right < left)
        {
            dfs(s + ")", left, right + 1, results, n);
        }
    }
};

int main()
{
    cout << "Running Generate_Paranthese.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/Backtracking/Generate_Paranthese/generate__paranthese.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        results = []

        def dfs(s, opened, closed):
            if len(s) == 2 * n:
                results.append(s)

            if opened < n:
                dfs(s + "(", opened + 1, closed)

            if closed < opened:
                dfs(s + ")", opened, closed + 1)

        dfs("", 0, 0)

        return results


if __name__ == "__main__":
    sol = Solution()
    for i in range(5):
        print(sol.generateParenthesis(i))
    print("Running Solution...")
